rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,538 @@
+impl Meshable for Direction2d {}
+
+struct Ray2d {
+  point: Point2d, 
+  direction: Direction2d
+}
+impl Meshable for Ray2d {}
+
+struct Line2d {
+  point: Point2d, 
+  direction: Direction2d,
+}
+impl Meshable for Line2d {}
+impl Collider2d for Line2d {}
+
+struct LineSegment2d { 
+  start: Point2d, 
+  end: Point2d,
+}
+impl Meshable for LineSegment2d {}
+impl Collider2d for LineSegment2d {}
+
+struct PolyLine2d<const N: usize>{
+  points: [Point2d; N],
+}
+impl Meshable for PolyLine2d {}
+impl Collider2d for PolyLine2d {}
+
+struct Triangle2d([Point2d; 3]);
+impl Meshable for Triangle2d {}
+impl Collider2d for Triangle2d {}
+
+struct Quad2d([Point2d; 4]);
+impl Meshable for Quad2d {}
+impl Collider2d for Quad2d {}
+
+/// A regular polygon, such as a square or hexagon.
+struct RegularPolygon2d {
+  /// The circumcircle that all points of the regular polygon lie on.
+  circumcircle: Circle2d,
+  /// Number of faces.
+  faces: u8,
+  /// Clockwise rotation of the polygon about the origin. At zero rotation, a point will always be located at the 12 o'clock position.
+  orientation: Angle,
+}
+impl Meshable for RegularPolygon2d {}
+impl Collider2d for RegularPolygon2d {}
+ 
+struct Polygon2d <const N: usize>{
+  points: [Point; N],
+}
+impl Meshable for Polygon2d {}
+impl Collider2d for Polygon2d {}
+
+/// Circle types
+
+struct Circle2d {
+  radius: f32,
+}
+impl Meshable for Circle2d {}
+
+struct BoundingCircle2d {
+  circle: Circle2d,
+  translation: Vec2,
+}
+impl Meshable for BoundingCircle2d {}
+impl Bounding2d for BoundingCircle2d {}
+
+struct CircleCollider2d {
+  sphere: Circle2d,
+  translation: Vec2,
+}
+impl Meshable for CircleCollider2d {}
+impl Collider2d for CircleCollider2d {}
+
+// Box Types
+
+struct Box2d {
+  half_extents: Vec2,
+}
+impl Meshable for Box2d
+
+struct BoundingBox2d {
+  box: Box2d,
+  translation: Vec2,
+}
+impl Meshable for BoundingBox2d {}
+impl Bounding2d for BoundingBox2d {}
+type Aabb2d = BoundingBox2d;
+
+struct BoxCollider2d {
+  box: Box2d,
+  translation: Vec2,
+  rotation: Mat2,
+}
+impl Meshable for BoxCollider2d {}
+impl Collider2d for BoxCollider2d {}
+type Obb2d = BoxCollider2d;
+
+// Capsule Types
+
+struct Capsule2d {
+  // Height of the cylindrical section
+  height: f32,
+  radius: f32,
+}
+impl Meshable for Capsule2d {}
+
+struct CapsuleCollider2d {
+  capsule: Capsule2d,
+  translation: Vec2,
+  rotation: Mat2,
+}
+impl Meshable for CapsuleCollider2d {}
+impl Collider2d for CapsuleCollider2d {}
+
+```
+### Lack of `Transform`s
+
+Primitives colliders and bounding volumes are fully defined in space, and do not use `Transform` or `GlobalTransform`. This is an intentional decision. Because transforms can include nonuniform scale, they are fundamentally incompatible with shape primitives. We could use transforms in the future if the `translation`, `rotation`, and `scale` fields were distinct components, and shape primitives could be bundled with a `translation` or `rotation` if applicable.
```

Comment:
I think if we can figure out an ergonomic way to instantiate these, maybe having Bundle constructors, so one could do `TransformBundle::new(translation, rotation, scale)` and `TransformBundle::new_translation(translation)` or something like that?

Reply:
Yup, I agree these would be really helpful. However, I'm marking out of scope.